client/network/src/protocol/generic_proto/upgrade/notifications.rs

```diff
@@ -189,8 +186,10 @@ where TSubstream: AsyncRead + AsyncWrite + Unpin,
 			match mem::replace(this.handshake, NotificationsInSubstreamHandshake::Sent) {
 				NotificationsInSubstreamHandshake::Sent =>
 					return Stream::poll_next(this.socket.as_mut(), cx),
-				NotificationsInSubstreamHandshake::NotSent =>
-					return Poll::Pending,
+				NotificationsInSubstreamHandshake::NotSent => {
+					*this.handshake = NotificationsInSubstreamHandshake::NotSent;
```

Comment:
The main bugfix is here.
The rest of the PR is only the log statements that I added a bit everywhere in my attempts to understand the issue.